server/src/main/java/io/druid/segment/realtime/firehose/ReplayableFirehoseFactory.java

```diff
@@ -0,0 +1,313 @@
+      @JsonProperty("maxTempFileSize") Integer maxTempFileSize,
+      @JsonProperty("readFirehoseRetries") Integer readFirehoseRetries
+  )
+  {
+    Preconditions.checkNotNull(delegateFactory, "delegate cannot be null");
+    Preconditions.checkArgument(
+        !(delegateFactory instanceof ReplayableFirehoseFactory),
+        "Refusing to wrap another ReplayableFirehoseFactory"
+    );
+
+    this.delegateFactory = delegateFactory;
+    this.reportParseExceptions = reportParseExceptions == null
+                                 ? DEFAULT_REPORT_PARSE_EXCEPTIONS
+                                 : reportParseExceptions;
+    this.maxTempFileSize = maxTempFileSize == null ? DEFAULT_MAX_TEMP_FILE_SIZE : maxTempFileSize;
+    this.readFirehoseRetries = readFirehoseRetries == null ? DEFAULT_READ_FIREHOSE_RETRIES : readFirehoseRetries;
+
+    log.info(this.toString());
+  }
+
+  @Override
+  public Firehose connect(InputRowParser parser) throws IOException
+  {
+    if (firehose == null) {
+      firehose = new ReplayableFirehose(parser);
+    } else {
+      log.info("Rewinding and returning existing firehose");
+      firehose.rewind();
+    }
+
+    return firehose;
+  }
+
+  @Override
+  public void close() throws IOException
+  {
+    if (firehose != null) {
+      firehose.possiblyDeleteTempFiles();
+      firehose.close();
+    }
+
+    firehose = null;
+  }
+
+  public class ReplayableFirehose implements Firehose
+  {
+    private final JsonFactory factory = new DefaultObjectMapper(
+        new SmileFactory().enable(SmileGenerator.Feature.CHECK_SHARED_STRING_VALUES)
+    ).getFactory();
+
+    private final List<File> files = Lists.newArrayList();
+    private final List<String> dimensions;
+
+    private int fileIndex = 0;
+    private JsonParser jsonParser;
+    private Iterator<Row> it;
+
+    public ReplayableFirehose(InputRowParser parser) throws IOException
+    {
+      long counter = 0, totalBytes = 0, unparseable = 0, retryCount = 0;
+      Set<String> dimensionScratch = Sets.newHashSet();
+
+      File tmpDir = Files.createTempDir();
+      if (DELETE_TEMP_FILES) {
+        tmpDir.deleteOnExit();
+      }
+
+      long ingestStartMillis = System.currentTimeMillis();
+      do {
+        possiblyDeleteTempFiles();
+        try (Firehose delegateFirehose = delegateFactory.connect(parser)) {
+          while (delegateFirehose.hasMore()) {
+            File tmpFile = File.createTempFile("replayable-", null, tmpDir);
+            if (DELETE_TEMP_FILES) {
+              tmpFile.deleteOnExit();
+            }
+
+            files.add(tmpFile);
+            log.debug("Created file [%s]", tmpFile.getAbsolutePath());
+
+            try (CountingOutputStream cos = new CountingOutputStream(new FileOutputStream(tmpFile));
+                 JsonGenerator generator = factory.createGenerator(cos)) {
+
+              while (delegateFirehose.hasMore() && cos.getCount() < getMaxTempFileSize()) {
```

Comment:
what happens if the maxtempfilesize is incorrectly configured, such as a negative value?